### PR TITLE
Use [NotNullWhen(true)] in more places

### PR DESCRIPTION
**Pull request by stephentoub**
The issue fixed by https://github.com/dotnet/runtime/pull/47579 inspired me to search for other places the consuming experience could be improved by using `[NotNullWhen(true)]`, and found _a lot_.  Most of these are overrides of object.Equals, but some are TryParse methods, other Equals overloads, and a smattering of other APIs.  I essentially just did a quick regex search for methods that returned bool and took nullable reference type as the first input, and then added [NotNullWhen(true)] where it was obviously correct (I skipped over a few large areas, namely reflection, as this was already getting a bit out of hand).

For reference, the benefit of applying `[NotNullWhen(true)]` to a nullable argument is it lets the compiler know that if the method returns true then that argument wasn't null when the method was called, which lets you write code like:
```C#
if (string1.Equals(string2))
{
    Console.WriteLine(string2.Length); // no nullable warning because the compiler sees string2 is non-null
}
```

Note that for the most part I stayed away from modifying virtual methods, but obviously some of the overrides I modified aren't sealed, so _technically_ this could be a breaking change (if someone derives from one of those types and overrides Equals and has nullable enabled and doesn't have [NotNullWhen(true)] on the argument).  I can undo any folks feel queasy about.

cc: @jeffhandley, @krwq, @buyaa-n

### CODE REVIEW

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs [hunk 1]
```diff
@@ -164,7 +164,7 @@ internal void FreeHandle()
             FreeGCHandle();
         }
 
-        public override bool Equals(object? obj) => obj is JSObject other && JSHandle == other.JSHandle;
+        public override bool Equals([NotNullWhen(true)] object? obj) => obj is JSObject other && JSHandle == other.JSHandle;
```
Comment (buyaa-n) [inline src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs:167]: Seems forgot to import System.Diagnostics.CodeAnalysis

**Review by buyaa-n — APPROVED** [review]
Left a few comment, overall looks good to me

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs [hunk 2]
```diff
@@ -122,7 +122,7 @@ public enum ProtectionLevel
         public SslApplicationProtocol(byte[] protocol) { throw null; }
         public SslApplicationProtocol(string protocol) { throw null; }
         public System.ReadOnlyMemory<byte> Protocol { get { throw null; } }
-        public bool Equals(System.Net.Security.SslApplicationProtocol other) { throw null; }
+        public bool Equals([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] System.Net.Security.SslApplicationProtocol other) { throw null; }
```
Comment (krwq) [inline src/libraries/System.Net.Security/ref/System.Net.Security.cs:125]: I think you meant to put it on the line below

Comment (krwq) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs:125]: (might be worth considering to add analyzer for cases like this)

**Review by krwq — APPROVED** [review]
LGTM with minor comment

File: src/libraries/System.Net.Security/ref/System.Net.Security.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/ref/System.Net.Security.cs:125]: Oops, good catch, thanks.

File: src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSObject.cs:167]: Thanks

**Pull request merged**